File: ios/ColorArchive/Utils/ColorContrast.swift
import Foundation

enum WcagGrade: String, CaseIterable {
    case aaa = "AAA"
    case aa = "AA"
    case aaLarge = "AA Large"
    case fail = "Fail"

    var passed: Bool { self != .fail }

    static func from(ratio: Double) -> WcagGrade {
        if ratio >= 7.0 { return .aaa }
        if ratio >= 4.5 { return .aa }
        if ratio >= 3.0 { return .aaLarge }
        return .fail
    }
}

struct WcagContrastData {
    let vsWhite: Double
    let vsBlack: Double
    let whiteGrade: WcagGrade
    let blackGrade: WcagGrade
}

struct WcagPairing {
    let color: ColorRecord
    let ratio: Double
    let grade: WcagGrade
}

enum ColorContrast {

    static func contrastRatio(_ color1: ColorRecord, _ color2: ColorRecord) -> Double {
        let rgb1 = ColorConvert.hslToRgb(hue: color1.hue, saturation: color1.saturation, lightness: color1.lightness)
        let rgb2 = ColorConvert.hslToRgb(hue: color2.hue, saturation: color2.saturation, lightness: color2.lightness)
        let lum1 = ColorConvert.relativeLuminance(r: rgb1.r, g: rgb1.g, b: rgb1.b)
        let lum2 = ColorConvert.relativeLuminance(r: rgb2.r, g: rgb2.g, b: rgb2.b)
        let lighter = max(lum1, lum2)
        let darker = min(lum1, lum2)
        return ((lighter + 0.05) / (darker + 0.05) * 10).rounded() / 10
    }

    static func wcagContrast(hue: Int, saturation: Int, lightness: Int) -> WcagContrastData {
        let rgb = ColorConvert.hslToRgb(hue: hue, saturation: saturation, lightness: lightness)
        let lum = ColorConvert.relativeLuminance(r: rgb.r, g: rgb.g, b: rgb.b)
        let vsWhite = ((1 + 0.05) / (lum + 0.05) * 10).rounded() / 10
        let vsBlack = ((lum + 0.05) / (0 + 0.05) * 10).rounded() / 10
        return WcagContrastData(
            vsWhite: vsWhite,
            vsBlack: vsBlack,
            whiteGrade: WcagGrade.from(ratio: vsWhite),
            blackGrade: WcagGrade.from(ratio: vsBlack)
        )
    }

    static func wcagPairings(
        _ colors: [ColorRecord],
        base: ColorRecord,
        limit: Int = 8
    ) -> [WcagPairing] {
        var pairings: [WcagPairing] = []
        for candidate in colors where candidate.id != base.id {
            let ratio = contrastRatio(base, candidate)
            guard ratio >= 3 else { continue }
            let grade: WcagGrade = ratio >= 7 ? .aaa : ratio >= 4.5 ? .aa : .aaLarge
            pairings.append(WcagPairing(color: candidate, ratio: ratio, grade: grade))
        }
        pairings.sort { a, b in
            if a.grade != b.grade {
                let order: [WcagGrade: Int] = [.aaa: 0, .aa: 1, .aaLarge: 2, .fail: 3]
                return (order[a.grade] ?? 3) < (order[b.grade] ?? 3)
            }
            let aDist = ColorRelationships.hueDistance(a.color.hue, base.hue)
            let bDist = ColorRelationships.hueDistance(b.color.hue, base.hue)
            return aDist > bDist
        }
        return Array(pairings.prefix(limit))
    }
}

File: src/lib/color-contrast.ts
import type { ColorRecord } from "@/src/types/color";
import { hslToRgb } from "./color-convert";
import { getHueDistance } from "./color-relationships";

export interface WcagContrastData {
  vsWhite: number;
  vsBlack: number;
  whiteGrade: "AA" | "AA Large" | "Fail";
  blackGrade: "AA" | "AA Large" | "Fail";
}

export interface WcagPairing {
  color: ColorRecord;
  ratio: number;
  grade: "AAA" | "AA" | "AA Large";
}

export function getRelativeLuminance(r: number, g: number, b: number): number {
  const toLinear = (c: number) => {
    const sRGB = c / 255;
    return sRGB <= 0.04045 ? sRGB / 12.92 : Math.pow((sRGB + 0.055) / 1.055, 2.4);
  };
  return 0.2126 * toLinear(r) + 0.7152 * toLinear(g) + 0.0722 * toLinear(b);
}

function wcagGrade(ratio: number): "AA" | "AA Large" | "Fail" {
  if (ratio >= 4.5) return "AA";
  if (ratio >= 3) return "AA Large";
  return "Fail";
}

/**
 * Should text on this colour be dark rather than light?
 *
 * One function so two surfaces cannot disagree. The colour detail hero and the
 * sticky bar each decided this for themselves and differed on 21 of the 5,446
 * archive colours: the hero compared getWcagContrast's ratios, which are rounded
 * to one decimal before they are returned, so a pair like 4.55 vs 4.54 came back
 * as 4.6 vs 4.5 in one place and was compared unrounded in the other. Rounding
 * belongs in display, not in a branch.
 */
export function prefersDarkText(r: number, g: number, b: number): boolean {
  const luminance = getRelativeLuminance(r, g, b);
  const vsWhite = 1.05 / (luminance + 0.05);
  const vsBlack = (luminance + 0.05) / 0.05;
  return vsBlack > vsWhite;
}

export function getWcagContrast(hue: number, saturation: number, lightness: number): WcagContrastData {
  const { r, g, b } = hslToRgb(hue, saturation, lightness);
  const colorLum = getRelativeLuminance(r, g, b);
  const vsWhite = (1 + 0.05) / (colorLum + 0.05);
  const vsBlack = (colorLum + 0.05) / (0 + 0.05);
  return {
    vsWhite: Math.round(vsWhite * 10) / 10,
    vsBlack: Math.round(vsBlack * 10) / 10,
    whiteGrade: wcagGrade(vsWhite),
    blackGrade: wcagGrade(vsBlack),
  };
}

export function getContrastRatio(color1: ColorRecord, color2: ColorRecord): number {
  const rgb1 = hslToRgb(color1.hue, color1.saturation, color1.lightness);
  const rgb2 = hslToRgb(color2.hue, color2.saturation, color2.lightness);
  const lum1 = getRelativeLuminance(rgb1.r, rgb1.g, rgb1.b);
  const lum2 = getRelativeLuminance(rgb2.r, rgb2.g, rgb2.b);
  const lighter = Math.max(lum1, lum2);
  const darker = Math.min(lum1, lum2);
  return Math.round(((lighter + 0.05) / (darker + 0.05)) * 10) / 10;
}

export function getWcagPairings(
  colors: readonly ColorRecord[],
  baseColor: ColorRecord,
  limit = 8,
): WcagPairing[] {
  const pairings: WcagPairing[] = [];

  for (const candidate of colors) {
    if (candidate.id === baseColor.id) continue;
    const ratio = getContrastRatio(baseColor, candidate);
    if (ratio >= 3) {
      const grade = ratio >= 7 ? "AAA" as const : ratio >= 4.5 ? "AA" as const : "AA Large" as const;
      pairings.push({ color: candidate, ratio, grade });
    }
  }

  // Sort: AAA first, then AA, then AA Large. Within same grade, prefer visual diversity
  pairings.sort((a, b) => {
    if (a.grade !== b.grade) {
      const gradeOrder = { "AAA": 0, "AA": 1, "AA Large": 2 };
      return gradeOrder[a.grade] - gradeOrder[b.grade];
    }
    // Within same grade, prefer colors from different families (diversity)
    const aDist = getHueDistance(a.color.hue, baseColor.hue);
    const bDist = getHueDistance(b.color.hue, baseColor.hue);
    return bDist - aDist; // farther hues first for diversity
  });

  return pairings.slice(0, limit);
}
